File: src/lexer.py
"""
Lexer for BASIC source code.
"""
from typing import List

from .tokens import Token, TokenType, KEYWORDS


class LexerError(Exception):
    def __init__(self, message: str, line: int, column: int):
        self.message = message
        self.line = line
        self.column = column
        super().__init__(f"{message} at line {line}, column {column}")


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1
        self.column = 1
        self.line_start = 0

    def _current(self) -> str:
        if self.pos >= len(self.source):
            return "\0"
        return self.source[self.pos]

    def _peek(self, offset: int = 1) -> str:
        p = self.pos + offset
        if p >= len(self.source):
            return "\0"
        return self.source[p]

    def _advance(self) -> str:
        if self.pos >= len(self.source):
            return "\0"
        c = self.source[self.pos]
        self.pos += 1
        if c == "\n":
            self.line += 1
            self.column = 1
            self.line_start = self.pos
        else:
            self.column += 1
        return c

    def _skip_whitespace(self) -> None:
        while self._current() in " \t\r":
            self._advance()

    def _read_number(self) -> Token:
        start_line, start_col = self.line, self.column
        buf = []
        while self._current().isdigit():
            buf.append(self._advance())
        if self._current() == "." and self._peek().isdigit():
            buf.append(self._advance())
            while self._current().isdigit():
                buf.append(self._advance())
            value = float("".join(buf))
        else:
            value = int("".join(buf))
        return Token(TokenType.NUMBER, value, start_line, start_col)

    def _read_string(self) -> Token:
        start_line, start_col = self.line, self.column
        quote = self._current()
        if quote not in '"\'':
            raise LexerError("Expected string quote", self.line, self.column)
        self._advance()
        buf = []
        while self._current() != quote and self._current() != "\0":
            if self._current() == "\n":
                raise LexerError("Unterminated string", self.line, self.column)
            if self._current() == "\\":
                self._advance()
                c = self._current()
                if c == "n":
                    buf.append("\n")
                elif c == "t":
                    buf.append("\t")
                elif c == quote:
                    buf.append(quote)
                else:
                    buf.append(c)
                self._advance()
            else:
                buf.append(self._advance())
        if self._current() != quote:
            raise LexerError("Unterminated string", self.line, self.column)
        self._advance()
        return Token(TokenType.STRING, "".join(buf), start_line, start_col)

    def _read_ident_or_keyword(self) -> Token:
        start_line, start_col = self.line, self.column
        buf = []
        c = self._current()
        if not (c.isalpha() or c == "_"):
            raise LexerError("Expected identifier", self.line, self.column)
        while c.isalnum() or c == "_":
            buf.append(self._advance())
            c = self._current()
        name = "".join(buf).upper()
        token_type = KEYWORDS.get(name, TokenType.IDENT)
        value = name if token_type == TokenType.IDENT else None
        if token_type == TokenType.IDENT:
            value = "".join(buf)  # preserve original case for variable names in some BASICs; we use upper
        return Token(token_type, value or name, start_line, start_col)

    def _read_rem(self) -> Token:
        start_line, start_col = self.line, self.column
        # REM has been consumed; rest of line is comment text
        buf = []
        while self._current() != "\n" and self._current() != "\0":
            buf.append(self._advance())
        return Token(TokenType.REM, "".join(buf).strip(), start_line, start_col)

    def _next_token(self) -> Token:
        self._skip_whitespace()
        start_line, start_col = self.line, self.column
        c = self._current()

        if c == "\0":
            return Token(TokenType.EOF, None, start_line, start_col)

        if c == "\n":
            self._advance()
            return Token(TokenType.NEWLINE, None, start_line, start_col)

        if c == ":":
            self._advance()
            return Token(TokenType.COLON, ":", start_line, start_col)

        if c == '"' or c == "'":
            return self._read_string()

        if c.isdigit():
            return self._read_number()

        if c.isalpha() or c == "_":
            # Check for REM to consume rest of line
            start = self.pos
            buf = []
            while self._current().isalnum() or self._current() == "_":
                buf.append(self._advance())
            name = "".join(buf).upper()
            if name == "REM":
                return self._read_rem()
            # Put back and use normal ident/keyword
            self.pos = start
            self.line, self.column = start_line, start_col
            return self._read_ident_or_keyword()

        # Two-char operators
        if c == "<":
            self._advance()
            if self._current() == ">":
                self._advance()
                return Token(TokenType.NE, "<>", start_line, start_col)
            if self._current() == "=":
                self._advance()
                return Token(TokenType.LE, "<=", start_line, start_col)
            return Token(TokenType.LT, "<", start_line, start_col)
        if c == ">":
            self._advance()
            if self._current() == "=":
                self._advance()
                return Token(TokenType.GE, ">=", start_line, start_col)
            return Token(TokenType.GT, ">", start_line, start_col)

        # Single-char
        if c == "+":
            self._advance()
            return Token(TokenType.PLUS, "+", start_line, start_col)
        if c == "-":
            self._advance()
            return Token(TokenType.MINUS, "-", start_line, start_col)
        if c == "*":
            self._advance()
            return Token(TokenType.STAR, "*", start_line, start_col)
        if c == "/":
            self._advance()
            return Token(TokenType.SLASH, "/", start_line, start_col)
        if c == "=":
            self._advance()
            return Token(TokenType.EQ, "=", start_line, start_col)
        if c == "(":
            self._advance()
            return Token(TokenType.LPAREN, "(", start_line, start_col)
        if c == ")":
            self._advance()
            return Token(TokenType.RPAREN, ")", start_line, start_col)
        if c == ",":
            self._advance()
            return Token(TokenType.COMMA, ",", start_line, start_col)

        raise LexerError(f"Unexpected character: {c!r}", self.line, self.column)

    def tokenize(self) -> List[Token]:
        tokens = []
        while True:
            t = self._next_token()
            tokens.append(t)
            if t.type == TokenType.EOF:
                break
        return tokens


def tokenize(source: str) -> List[Token]:
    return Lexer(source).tokenize()

File: webapp/src/compiler/lexer.ts
import type { Token } from './tokens';
import { TokenType, KEYWORDS } from './tokens';

export class LexerError extends Error {
  line: number;
  column: number;
  constructor(message: string, line: number, column: number) {
    super(`${message} at line ${line}, column ${column}`);
    this.name = 'LexerError';
    this.line = line;
    this.column = column;
  }
}

export function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let pos = 0, line = 1, column = 1;

  function current(): string {
    return pos >= source.length ? '\0' : source[pos];
  }
  function peek(off = 1): string {
    const p = pos + off;
    return p >= source.length ? '\0' : source[p];
  }
  function advance(): string {
    if (pos >= source.length) return '\0';
    const c = source[pos++];
    if (c === '\n') { line++; column = 1; } else { column++; }
    return c;
  }
  function skipWhitespace(): void {
    while (current() === ' ' || current() === '\t' || current() === '\r') advance();
  }
  function readNumber(): Token {
    const startLine = line, startCol = column;
    const buf: string[] = [];
    while (/\d/.test(current())) buf.push(advance());
    if (current() === '.' && /\d/.test(peek())) {
      buf.push(advance());
      while (/\d/.test(current())) buf.push(advance());
      return { type: TokenType.NUMBER, value: parseFloat(buf.join('')), line: startLine, column: startCol };
    }
    return { type: TokenType.NUMBER, value: parseInt(buf.join(''), 10), line: startLine, column: startCol };
  }
  function readString(): Token {
    const startLine = line, startCol = column;
    const quote = current();
    if (quote !== '"' && quote !== "'") throw new LexerError('Expected string quote', line, column);
    advance();
    const buf: string[] = [];
    while (current() !== quote && current() !== '\0') {
      if (current() === '\n') throw new LexerError('Unterminated string', line, column);
      if (current() === '\\') {
        advance();
        const c = current();
        if (c === 'n') buf.push('\n');
        else if (c === 't') buf.push('\t');
        else if (c === quote) buf.push(quote);
        else buf.push(c);
        advance();
      } else buf.push(advance());
    }
    if (current() !== quote) throw new LexerError('Unterminated string', line, column);
    advance();
    return { type: TokenType.STRING, value: buf.join(''), line: startLine, column: startCol };
  }
  function readIdentOrKeyword(): Token {
    const startLine = line, startCol = column;
    const buf: string[] = [];
    if (!/[a-zA-Z_]/.test(current())) throw new LexerError('Expected identifier', line, column);
    while (/[a-zA-Z0-9_]/.test(current())) buf.push(advance());
    const name = buf.join('').toUpperCase();
    if (name === 'REM') {
      const rest: string[] = [];
      while (current() !== '\n' && current() !== '\0') rest.push(advance());
      return { type: TokenType.REM, value: rest.join('').trim(), line: startLine, column: startCol };
    }
    const tokenType = KEYWORDS[name] ?? TokenType.IDENT;
    const value = tokenType === TokenType.IDENT ? buf.join('') : name;
    return { type: tokenType, value, line: startLine, column: startCol };
  }
  function nextToken(): Token {
    skipWhitespace();
    const startLine = line, startCol = column;
    const c = current();
    if (c === '\0') return { type: TokenType.EOF, value: null, line: startLine, column: startCol };
    if (c === '\n') { advance(); return { type: TokenType.NEWLINE, value: null, line: startLine, column: startCol }; }
    if (c === ':') { advance(); return { type: TokenType.COLON, value: ':', line: startLine, column: startCol }; }
    if (c === '"' || c === "'") return readString();
    if (/\d/.test(c)) return readNumber();
    if (/[a-zA-Z_]/.test(c)) return readIdentOrKeyword();
    if (c === '<') {
      advance();
      if (current() === '>') { advance(); return { type: TokenType.NE, value: '<>', line: startLine, column: startCol }; }
      if (current() === '=') { advance(); return { type: TokenType.LE, value: '<=', line: startLine, column: startCol }; }
      return { type: TokenType.LT, value: '<', line: startLine, column: startCol };
    }
    if (c === '>') {
      advance();
      if (current() === '=') { advance(); return { type: TokenType.GE, value: '>=', line: startLine, column: startCol }; }
      return { type: TokenType.GT, value: '>', line: startLine, column: startCol };
    }
    const single: [string, (typeof TokenType)[keyof typeof TokenType]][] = [
      ['+', TokenType.PLUS], ['-', TokenType.MINUS], ['*', TokenType.STAR], ['/', TokenType.SLASH],
      ['=', TokenType.EQ], ['(', TokenType.LPAREN], [')', TokenType.RPAREN], [',', TokenType.COMMA]
    ];
    for (const [ch, tt] of single) {
      if (c === ch) { advance(); return { type: tt, value: ch, line: startLine, column: startCol }; }
    }
    throw new LexerError(`Unexpected character: ${c}`, line, column);
  }

  for (;;) {
    const t = nextToken();
    tokens.push(t);
    if (t.type === TokenType.EOF) break;
  }
  return tokens;
}
